xcdat/dataset.py

```diff
@@ -779,20 +633,22 @@ def _get_cftime_coords(
     np.ndarray
         An array of `cftime` coordinates.
     """
-    # Starting from the reference date, create an array of `datetime` objects
-    # by adding each offset (a numerically encoded value) to the reference date.
-    # The `parse.parse` default is set to datetime(2000, 1, 1), with each
-    # component being a placeholder if the value does not exist. For example, 1
-    # and 1 are placeholders for month and day if those values don't exist.
+    # Create an array of `datetime` objects by adding the number offset to the
+    # reference date. The `parse.parse` default is set to datetime(2000, 1, 1),
+    # with each component being a placeholder if the value does not exist.
     ref_datetime: datetime = parser.parse(ref_date, default=datetime(2000, 1, 1))
-    offsets = np.array(
+
+    # Convert offsets to `np.float64` to avoid "TypeError: unsupported type for
+    # timedelta days component: numpy.int64".
+    offsets = offsets.astype("float64")
+    times = np.array(
         [ref_datetime + rd.relativedelta(**{units: offset}) for offset in offsets],
         dtype="object",
     )
 
     # Convert the array of `datetime` objects into `cftime` objects based on
     # the calendar type.
     date_type = get_date_type(calendar)
```

Comment:
A very common issue with the current PR is that datasets fail to open because of leap day issues:

```
import xcdat as xc

ds = xc.open_mfdataset(dpath)  
```

> File ~/code/xcdat/robust_tests/../xcdat/dataset.py:653, in _get_cftime_coords(ref_date, offsets, calendar, units)
>     650 # Convert the array of `datetime` objects into `cftime` objects based on
>     651 # the calendar type.
>     652 date_type = get_date_type(calendar)
> --> 653 coords = convert_times(times, date_type=date_type)

> ValueError: Cannot convert date 1916-02-29 12:00:00 to a date in the noleap calendar.  Reason: invalid day number provided in cftime.DatetimeNoLeap(1916, 2, 29, 12, 0, 0, 0, has_year_zero=True).

This suggestion seems to address that problem by using an xarray function to decode the time coordinate, but it is super slow (somehow it is even slower than `xr.open_mfdataset()` which does the same decoding):

```%timeit ds = xc.open_mfdataset(dpath); ds.close()```
> 3.26 s ± 4.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)                                                                                    

```%timeit ds = xc.open_mfdataset(dpath, decode_times=False); ds.close()```
> 865 ms ± 1.42 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)                                                                                   

```%timeit ds = xr.open_mfdataset(dpath); ds.close()```
> 909 ms ± 21.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)   

So there must be a better way to do this (and this suggestion should be rejected).

Reply:
> (somehow it is even slower than `xr.open_mfdataset()` which does the same decoding):

> `%timeit ds = xr.open_mfdataset(dpath); ds.close()`
> 
> > 909 ms ± 21.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

By default, `xr.open_mfdataset()` is not actually performing the same decoding as `xcdat.open_mfdataset()`.

You didn't call  `ds = xr.open_mfdataset(dpath, use_cftime=True)`, so xarray is decoding as `datetime` rather than `cftime` objects.

As noted in our Slack conversation, xarray calls `cftime.num2date()` under the hood to decode time coordinates as `cftime`.  `cftime.num2date()` performance depends on the calendar type based on the docstring below.

```python
    For standard (Gregorian) calendars, this function uses vectorized
    operations, which makes it much faster than cftime.num2date. In such a
    case, the returned array will be of type np.datetime64.
```
Source: https://github.com/pydata/xarray/blob/7379923de756a2bcc59044d548f8ab7a68b91d4e/xarray/coding/times.py

**I think this explains the large performance differences, but we will need to test again.

Reply:
I edited my comment above a million times so sorry if you keep seeing updates.

Reply:
No problem (I didn't see a ton of updates)!

I don't think this is a vectorization issue, because the performance gap is still there if I use `use_cftime=True`:

```
fn = '/p/css03/cmip5_css02/data/cmip5/output1/INM/inmcm4/historical/mon/atmos/Amon/r1i1p1/tas/1/tas_Amon_inmcm4_historical_r1i1p1_185001-200512.nc'
%timeit ds = xc.open_dataset(fn); ds.close()                                                                                                 
1.71 s ± 1.69 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)                                                                                                                                       
                                                                                                                                                      
%timeit ds = xr.open_dataset(fn, use_cftime=True); ds.close()                                                                                
858 ms ± 8.63 ms per loop (mean ± std. dev. of 7 runs, 1 loop each) 
```

It [creates some classes to decode time](https://github.com/pydata/xarray/blob/7379923de756a2bcc59044d548f8ab7a68b91d4e/xarray/coding/times.py#L711), which may be executed on demand (after `open_dataset`)?

Reply:
_TLDR - We concluded in our debugging meeting that we probably don't need to worry about performance._

* Xarray seems to be deferring decoding times and time bounds through lazy operations (refer to complete traceback below). As a result, decoding only occurs when necessary (e.g., computation, array manipulation). **Xarray takes the decoding runtime performance hit later.**
* xCDAT decodes times and time bounds immediately after opening a dataset, which are eager operations. We need to do this in order to handle decoding both CF and non-CF compliant time units. **xCDAT takes the decoding runtime performance hit upfront.**
---

Here's the complete traceback where we figured out that xarray is lazily decoding time and time bounds:
1. Each backend has a subclass for [BackendEntryPoint](https://github.com/pydata/xarray/blob/7379923de756a2bcc59044d548f8ab7a68b91d4e/xarray/backends/common.py#L361)
   * For example, opening a dataset with `netcdf4` results in a call to [`NetCDF4BackendEntrypoint`](https://github.com/pydata/xarray/blob/7379923de756a2bcc59044d548f8ab7a68b91d4e/xarray/backends/netCDF4_.py#L518)
3. Each backend subclass has its own `open_dataset()` method
   * Example [`NetCDF4BackendEntrypoint.open_dataset()`](https://github.com/pydata/xarray/blob/7379923de756a2bcc59044d548f8ab7a68b91d4e/xarray/backends/netCDF4_.py#L534)
4. Each of these `open_dataset()` methods have a call to `StoreBackendEntrypoint().open_dataset()`
    * Method definition: https://github.com/pydata/xarray/blob/7379923de756a2bcc59044d548f8ab7a68b91d4e/xarray/backends/netCDF4_.py#L567
5. Within `StoreBackendEntrypoint().open_dataset()`, there is a call to `decode_cf_variables()`
   * Function definition: https://github.com/pydata/xarray/blob/7379923de756a2bcc59044d548f8ab7a68b91d4e/xarray/backends/store.py#L29-L39
6. Within `decode_cf_variables()`, there is a call to `decode_cf_variable()`
   * Function definition: https://github.com/pydata/xarray/blob/7379923de756a2bcc59044d548f8ab7a68b91d4e/xarray/conventions.py#L522-L532
7. Within `decode_cf_variable()`, there is a call to `times.CFTimedeltaCoder().decode()` and `times.CFDatetimeCoder().decode()`
8. Within `times.CFTimedeltaCoder().decode()` and `times.CFDatetimeCoder.decode()`, there are calls to `lazy_elemwise_func()`
   * https://github.com/pydata/xarray/blob/7379923de756a2bcc59044d548f8ab7a68b91d4e/xarray/coding/times.py#L729
   * https://github.com/pydata/xarray/blob/7379923de756a2bcc59044d548f8ab7a68b91d4e/xarray/coding/times.py#L706
 
Finally, `lazy_elem_wise_func()` is defined as: 
```python
def lazy_elemwise_func(array, func, dtype):
    """Lazily apply an element-wise function to an array.
    Parameters
    ----------
    array : any valid value of Variable._data
    func : callable
        Function to apply to indexed slices of an array. For use with dask,
        this should be a pickle-able object.
    dtype : coercible to np.dtype
        Dtype for the result of this function.

    Returns
    -------
    Either a dask.array.Array or _ElementwiseFunctionArray.
    """
    if is_duck_dask_array(array):
        import dask.array as da


        return da.map_blocks(func, array, dtype=dtype)
    else:
        return _ElementwiseFunctionArray(array, func, dtype)
```
_Source: https://github.com/pydata/xarray/blob/7379923de756a2bcc59044d548f8ab7a68b91d4e/xarray/coding/variables.py#L80-L100_

With a return statement calling:
```python
class _ElementwiseFunctionArray(indexing.ExplicitlyIndexedNDArrayMixin):
    """Lazily computed array holding values of elemwise-function.

    Do not construct this object directly: call lazy_elemwise_func instead.

    Values are computed upon indexing or coercion to a NumPy array.
    """


    def __init__(self, array, func, dtype):
        assert not is_duck_dask_array(array)
        self.array = indexing.as_indexable(array)
        self.func = func
        self._dtype = dtype


    @property
    def dtype(self):
        return np.dtype(self._dtype)


    def __getitem__(self, key):
        return type(self)(self.array[key], self.func, self.dtype)


    def __array__(self, dtype=None):
        return self.func(self.array)


    def __repr__(self):
        return "{}({!r}, func={!r}, dtype={!r})".format(
            type(self).__name__, self.array, self.func, self.dtype
        )
```
_Source: https://github.com/pydata/xarray/blob/7379923de756a2bcc59044d548f8ab7a68b91d4e/xarray/coding/variables.py#L50-L77_